fairlearn/metrics/_metric_frame.py

```diff
@@ -117,24 +126,127 @@ def __init__(self,
                  y_pred, *,
                  sensitive_features,
                  control_features: Optional = None,
-                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None):
+                 sample_params: Optional[Union[Dict[str, Any], Dict[str, Dict[str, Any]]]] = None,
+                 streaming=False):
         """Read a placeholder comment."""
+        if len(y_true) == 0 and not streaming:
+            raise ValueError(_BAD_INIT_ERROR)
+
+        self._y_t = None
+        self._y_p = None
+        self._s_f = sensitive_features
+        self._c_f = control_features
+        self._streaming = streaming
+        self.func_dict = self._process_functions(metric, sample_params)
+
+        self._cf_names = None
+        self._sf_names = None
+
+        if not streaming:
+            # Since we are not streaming, we can compute the metrics already.
+            # This keeps the existing behavior where we compute the metrics in the
+            # constructor.
+            check_consistent_length(y_true, y_pred)
+            y_t = _convert_to_ndarray_and_squeeze(y_true)
+            y_p = _convert_to_ndarray_and_squeeze(y_pred)
+            self._compute_metric(y_t, y_p, sensitive_features, control_features)
+        else:
+            # If we are streaming, we wait, but initialize accumulators.
+            assert y_true == [] and y_pred == [] and sensitive_features == [], _NON_EMPTY_INIT_STR
+            self._y_t = []
+            self._y_p = []
+            self._overall = None
+            self._by_group = None
+
+    def add_data(self,
+                 y_true,
+                 y_pred, *,
+                 sensitive_features,
+                 control_features: Optional = None,
+                 ):
```

Comment:
Do we want to add `sample_params` so as to pretty much follow the same format as constructor minus `metric`?

Reply:
I think this would lead to errors.
Example:
```
met = MetricFrame(skm.recall_score, y_true=[], y_pred=[], sensitive_features=[], streaming=True, sampling_params={'average':'micro'})

met_strt.add_data(..., sample_params={'average':'macro'})
....
met_strt.add_data(..., sample_params={'average':'micro'})
```

Then we compute the value differently for different batches.